torch/distributed/_state_dict_utils.py

```diff
@@ -305,28 +370,29 @@ def _create_cpu_state_dict(
     tensors can be placed on pin_memory or share_memory based on the provided arguments.
     """
 
-    if pin_memory and share_memory:
-        raise ValueError(
-            "Cannot allocate both memory on both pin_memory and share_memory"
-        )
-
     def tensor_func(
         obj: torch.Tensor,
         pg: Optional[dist.ProcessGroup],
         device: Optional[torch.device],
-        companion_obj: Any,
+        _: Any,
     ) -> torch.Tensor:
         if len(obj.size()) == 0:
             return torch.tensor(0, dtype=obj.dtype)
 
         if share_memory:
-            return torch.empty(
-                *tuple(companion_obj.size()), dtype=companion_obj.dtype
-            ).share_memory_()
+            t = torch.empty(*tuple(obj.size()), dtype=obj.dtype).share_memory_()
+            if pin_memory:
+                succ = torch.cuda.cudart().cudaHostRegister(
```

Comment:
is there any perf implication from creating a new `cudart()` object every call? lets figure out if this is lightweight or if we should cache this object

Reply:
This should be only called once as this API is to prepare the buffer for offloading.